### PR TITLE
1721: Update SAML to use new manager claims

### PR DESCRIPTION
**Pull request by turegjorup**
Updated plugin to use new `personaleLederUPN` claims

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,11 +14,11 @@ which in turn delegates the heavy lifting to `KimaiPlugin\AakSamlBundle\Service\
 
 - `Office (officeId)` -> `name`
 
-We map `Office` (e.g. "ITK Development") to a Kimai team. Kimai has a unique constraint on team names. We include the id
-to ensure uniqueness. E.g. "ITK Development (6530)"
+We map `Office` (e.g. "ITK Development") to a Kimai team. Kimai has a unique constraint on team names. We include the
+manger (personaleleder) email to ensure uniqueness. E.g. "ITK Development (<jane@examlpe.org>)"
```
Comment (rimi-itk) [inline README.md]:
```suggestion
manager ("personaleleder") email to ensure uniqueness. E.g. "ITK Development (<jane@examlpe.org>)"
```

File: Service/SamlDataHydrateService.php [hunk 2]
```diff
@@ -28,24 +28,36 @@ public function __construct(
...
             $this->hydrateTeamLeadForTeam($teamLeadTeam, $user);
-        } else {
-            $teamLead = $this->hydrateTeamLead($memberTeam, $samlDto);
-        }
 
-        $this->hydrateUser($user, $samlDto, $teamLead, $memberTeam);
+            foreach ($user->getTeams() as $team) {
+                if ($team->isTeamlead($user)) {
+                    foreach ($team->getUsers() as $teamMember) {
+                        if (!in_array($teamMember->getUsername(), $samlDto->employeeList) && $teamMember->getUsername() !== $user->getUsername()) {
```
Comment (rimi-itk) [inline Service/SamlDataHydrateService.php]: Under the assumption that comparing strings are faster than searching an array, a (very small) performance gain might be achieved by switching the expressions around:

```suggestion
                        if ($teamMember->getUsername() !== $user->getUsername() && !in_array($teamMember->getUsername(), $samlDto->employeeList)) {
```

File: Service/SamlDTO.php [hunk 3]
```diff
@@ -137,6 +144,11 @@ public function __construct(array $samlAttributes)
         $this->departmentId = $this->departmentIds[2] ?? null;
         $this->subDepartmentId = $this->departmentIds[3] ?? null;
         $this->officeId = $this->departmentIds[4] ?? null;
+
+        $employeeEmailArray = \explode(';', $this->getAttributeValue(self::EMPLOYEE_LIST_ATTRIBUTE, $samlAttributes));
+        $this->employeeList = \array_filter($employeeEmailArray, function ($value) {
+            return '' !== $value;
+        });
```
Comment (rimi-itk) [inline Service/SamlDTO.php]: I would use [`preg_split`](https://www.php.net/manual/en/function.preg-split.php) for this

```suggestion
        $this->employeeList = \preg_split('/\s*;\*/', $this->getAttributeValue(self::EMPLOYEE_LIST_ATTRIBUTE, $samlAttributes), -1, PREG_SPLIT_NO_EMPTY);
```

but that's a matter of habit and preference.

I would also use just `\array_filter($employeeEmailArray)` to weed out empty values, but being explicit makes sense.

However, your solution is perfectly fine and sound.

**Review by rimi-itk — APPROVED** [review]
Approved with some suggestions.

File: Service/SamlDataHydrateService.php (hunk 2 above)
Comment (turegjorup) [reply Service/SamlDataHydrateService.php]: Given that there can only be one team lead for each team (our business rules, not Kimai's) we will always have to do the array search N-1 times regardless of the order of expressions. Have refactored to use `isset()`

File: Service/SamlDTO.php (hunk 3 above)
Comment (turegjorup) [reply Service/SamlDTO.php]: Yes. But `explode()` is [faster](https://www.php.net/manual/en/function.preg-split.php). So if we are micro optimizing then `preg_split()` is the wrong choice. Only advantage is that `preg_split()` can filter empty values. Benchmarking which is faster, `array_filter(explode())` or `preg_split()` is left as an exercise for the reader 😄 

Note: Most array sized expected to be `N < 20`

**Pull request merged**